Adding test-cases for ipa-cacert-manage

amore17 opened the pull request:
Scenario1:	Setup external CA1 and install ipa-server with CA1.
     		Setup exteranal CA2 and renew ipa-server with CA2.
		Get information to compare CA change for ca1 and CA2
     		it should show different Issuer between install
		and renewal.

Scenario2:	Renew CA Cert on Replica using ipa-cacert-manage
		verify that replica is caRenewalMaster

Signed-off-by: Anuja More <amore@redhat.com>

slaykovsky commented:
Hello @amore17 

Can you please add a commit to your PR and add tests you've implemented into `.freeipa-pr-ci.yaml` file?

It would be good to remove all others tests from this file and add only ones you need to test. Also, don't delete `fedora27/build` task from there. Another suggestion is to increase timeout just in case if tests will take a lot of time.

Rezney commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +279,77 @@ def test_install_external_ca(self):
+
+class TestMultipleExternalCA(IntegrationTest):
+    """setup ext-ca1 install ipa-server with ca1
+     setup ext-ca2 renew ipa-server with ext-ca2
```
Sounds a little confusing. Could you re-phrase pls?

amore17 replied:
In commit it is changed to "Setup externally signed ca1
     install ipa-server with with externally signed ca1
     Setup externally signed ca2 and renew ipa-server with
     externally signed ca2 and check the difference in certificate"

Rezney commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +279,77 @@ def test_install_external_ca(self):
+        install_server_external_ca_step1(self.master)
+        # Sign CA, transport it to the host and get ipa a root ca paths.
+
+        test_dir1 = os.path.join(self.master.config.test_dir, 'CA1')
```
I would recommend to create temporary directory here e.g.:
`tempfile.mkdtemp(prefix="extCA1-")`

This way you do not need to do cleanup (which is missing here) later.

Applies to all other directories.

Rezney commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +279,77 @@ def test_install_external_ca(self):
+        install_server_external_ca_step2(self.master, ipa_ca_fname1,
+                                         root_ca_fname1)
+
+        str1 = "caSigningCert cert-pki-ca"
```
Please use a little bit descriptive variable names. E.g. instead of "str1" you can use "ca_nick".

Rezney commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +279,77 @@ def test_install_external_ca(self):
+                                         root_ca_fname1)
+
+        str1 = "caSigningCert cert-pki-ca"
+        info1 = self.master.run_command(['certutil', '-L', '-d',
```
Same here e.g. "certutil_out"

Rezney commented on ipatests/test_integration/test_replica_promotion.py:
```diff
@@ -484,6 +484,18 @@ def test_replica_not_marked_as_renewal_master(self):
+        self.assertCARenewalMaster(replica, replica.hostname)
+        # set master back to ca-renewal-master
+        master.run_command([paths.IPA_CACERT_MANAGE, 'renew'])
+        self.assertCARenewalMaster(master, master.hostname)
```
I would also check how other server (replica in this case) sees the situation. So:

`self.assertCARenewalMaster(master, master.hostname)`
`self.assertCARenewalMaster(replica, master.hostname)`

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        cert_out1 = self.master.run_command(['certutil', '-L', '-d',
+                                             paths.PKI_TOMCAT_ALIAS_DIR,
+                                             '-n', cert_nick]).stdout_text
+        assert "CN=RootCA1" in cert_out1
```
This block is a bit hard to read because the function arguments are heavily indented. The ``.stdout_text`` attribute access is easy to mess, too. Other test codes commonly assigns the output of ``run_command`` to a name ``result`` and then checks its ``stdout_text`` property.

```
        result = self.master.run_command([
            'certutil', '-L', '-d', paths.PKI_TOMCAT_ALIAS_DIR,
            '-n', "caSigningCert cert-pki-ca"
        ])
        assert "CN=RootCA1" in result.stdout_text
```

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        assert "CN=RootCA1" in cert_out1
+
+    def test_master_install_ca2(self):
+
```
Unnecessary new line

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        root_ca_fname2 = tempfile.mkdtemp(suffix='root_ca.crt', dir=paths.TMP)
+        ipa_ca_fname2 = tempfile.mkdtemp(suffix='ipa_ca.crt', dir=paths.TMP)
+
+        self.master.run_command([paths.IPA_CACERT_MANAGE, 'renew',
```
Try to make code more readable by minimizing indention levels and keeping arguments on line line:

```
        self.master.run_command([
            paths.IPA_CACERT_MANAGE, 'renew', '--external-ca'
        ])
```

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        ipa_csr = self.master.get_file_contents(paths.IPA_CA_CSR)
+
+        external_ca = ExternalCA()
+        # Create root CA
```
unnecessary comments like in the other function

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        cert_out2 = self.master.run_command(['certutil', '-L', '-d',
+                                             paths.PKI_TOMCAT_ALIAS_DIR,
+                                             '-n', cert_nick]).stdout_text
+        assert "CN=RootCA2" in cert_out2
```
Reindent both ``run_command`` calls and use ``result.stdout_text``

tiran commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,71 @@ def test_install_external_ca(self):
+        root_ca_fname1 = tempfile.mkdtemp(suffix='root_ca.crt', dir=paths.TMP)
+        ipa_ca_fname1 = tempfile.mkdtemp(suffix='ipa_ca.crt', dir=paths.TMP)
+
+        # Get IPA CSR as bytes
```
It's a good idea to separate logical blocks by comments or empty. But both new line and comment is a bit much and bloats the code. In most cases, a new line is a sufficient visual separator.

flo-renaud commented on ipatests/test_integration/test_external_ca.py:
```diff
@@ -279,3 +280,60 @@ def test_install_external_ca(self):
+class TestMultipleExternalCA(IntegrationTest):
+    """Setup externally signed ca1
+
+    install ipa-server with with externally signed ca1
```
Typo with with

tiran commented:
@amore17 please rebase your PR on master and fix the merge conflict.

tiran commented:
Github is still complaining about merge conflicts. Please squash all commits into a single commit and force-push your branch again.

amore17 commented:
@flo-renaud Thanks for your suggestion.
I added changes as per your review. And I removed changes from .freeipa-pr-ci.yaml 
@tiran  I resolved conflicts.

Rezney commented:
@tiran Sure, LGTM. Thanks all for your work here...
